### PR TITLE
Refactor Message._add_value and Message.from_dict a little bit

### LINKED ISSUE #464: Decrease complexity of oic.oauth2.message.Message._add_value
Part of  #311:
`src/oic/oauth2/message.py:378:1: C901 'Message._add_value' is too complex (37) [mccabe]`

I will separate the code into separate subfunctions based on type and/or refactor it a little bit.

The function is basically untested but private so I will probably try to write tests through `from_dict`.

The max-xomplexity can then be lowered to `32`.

### PR DESCRIPTION
**Pull request by tpazderka**
Decreasing complexity
Close #464

- [ ] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [ ] The documentation has been updated, if necessary.
---

### CODE REVIEW

**Comment by tpazderka** [comment]
Touch of refactoring and some more tests.

File: src/oic/oauth2/message.py [hunk 1]
```diff
@@ -470,6 +406,58 @@ def _add_value(self, skey, vtyp, key, val, _deser, null_allowed):
                 else:
                     self._dict[skey] = val
 
+    def _add_value_list(self, skey, vtype, key, val, _deser, null_allowed):
+        """Add value that is supposed to be a list.
```
Comment (lwm) [inline src/oic/oauth2/message.py]: Can you improve on this docstring? Not sure it gives a good overview.

**Comment by lwm** [comment]
In general, nice one for taking this on! I am wondering - could we use some early `return` statements, in the new `_add_value_list` to avoid the if/elif/else nesting? Each condition could then become an if, and return early when the value is set. There would be no need for a final else, in this case, we just raise an error. Just a thought, maybe it won't reab any better ;)

File: src/oic/oauth2/message.py (hunk 1 above)
Comment (tpazderka) [reply src/oic/oauth2/message.py]: Tried to improve, not sure if it is any better:)

**Comment by tpazderka** [comment]
I have tried the explicit `return` statements, let me know what you think. Personally it looks the same to me so I am not against that.

**Comment by lwm** [comment]
Ah! Nice, but I meant something more like:

``` diff
+        if isinstance(val, vtype):
+            if issubclass(vtype, Message):
+                self._dict[skey] = [val]
+                return
+
+            if _deser is not None:
+                try:
+                    self._dict[skey] = _deser(val, sformat="urlencoded")
+                    return
+                except Exception as exc:
+                    raise DecodeError(ERRTXT % (key, exc))
+
+            setattr(self, skey, [val])
+            return
+
+        if isinstance(val, list):
```

Was hoping it would reduce the complexity - however, writing it now, perhaps it is also a confusing read.

Up to you!

**Comment by tpazderka** [comment]
I was considering that but it looked more confusing to me. The decrease in complexity is probably no longer that important since we got it below the next offender in line.

If we figure out we need to refactor it even more later on, than we can do it then.

**Comment by tpazderka** [comment]
@lwm Can I get a review on this?

**Pull request merged**